### PR TITLE
Add support for explorer node/solve bootstrap deadlock issue

### PR DESCRIPTION
**Pull request by rlan35**
INFO [06-14|00:14:56.307] HOORAY!!!!!!! CONSENSUS REACHED!!!!!!!   port=9000  ip=127.0.0.1 myBlock=4 myViewID=3 phase=Announce mode=Normal funcName=github.com/harmony-one/harmony/consensus.(*Consensus).finalizeCommits           funcFile=consensus_v2.go      funcLine=574  numOfSignatures=0

INFO [06-14|00:15:03.017] [Explorer] Committing block into explorer DB port=19999 ip=127.0.0.1 msgBlock=1

### CODE REVIEW

File: cmd/harmony/main.go [hunk 1]
```diff
@@ -162,54 +164,60 @@ func initSetup() {
 		utils.BootNodes = bootNodeAddrs
 	}
 
-	ks = hmykey.GetHmyKeyStore()
+	fmt.Println(*isGenesis)
+	fmt.Println(*isExplorer)
+	if !*isExplorer { // Explorer node doesn't need the following setup
```
Comment (alajko) [inline cmd/harmony/main.go]: no need for print statement/debug code.

File: cmd/harmony/main.go [hunk 2]
```diff
@@ -427,6 +442,8 @@ func main() {
 	utils.SetLogContext(*port, *ip)
 	utils.SetLogVerbosity(log.Lvl(*verbosity))
 
+	fmt.Println(os.Args)
+	fmt.Println(*minPeers)
 	initSetup()
```
Comment (alajko) [inline cmd/harmony/main.go]: debug code.

File: cmd/harmony/main.go [hunk 3]
```diff
@@ -427,6 +442,8 @@ func main() {
...
 	nodeConfig := createGlobalConfig()
```
Comment (alajko) [inline cmd/harmony/main.go]: There are two many `if` conditions in this file,
 maybe we could simply have if `*isExplorer nodeConfig := createExplorerConfig() else nodeConfig := createGlobalConfig()` or we want to keep createGlobalConfig(), the if conditions in that function could be factored out into their own fucntions. e.g. `if *isExplorer nodeConfig := createExplorerConfig() else nodeConfig := createConsensusNodeConfig()` etc. Will make for better readability.

File: node/node_handler.go [hunk 4]
```diff
@@ -812,3 +823,64 @@ func (node *Node) ConsensusMessageHandler(msgPayload []byte) {
...
+		if err != nil {
+			utils.GetLogInstance().Debug("[Explorer] onAnnounce unable to parse msg", "error", err)
+			return
+		}
+		block := recvMsg.Payload
+
+		// check block header is valid
+		var blockObj types.Block
+		err = rlp.DecodeBytes(block, &blockObj)
+		node.Consensus.PbftLog.AddBlock(&blockObj)
```
Comment (chaosma) [inline node/node_handler.go]: have you cleaned the blockobj somewhere later?

Comment (rlan35) [reply node/node_handler.go]: Done.

File: cmd/harmony/main.go (hunk 2 above)
Comment (rlan35) [reply cmd/harmony/main.go]: Done.

File: cmd/harmony/main.go (hunk 1 above)
Comment (rlan35) [reply cmd/harmony/main.go]: Done.

**Comment by lzl124631x** [comment]
@rlan35 Ping. Possible to merge it these days?

File: cmd/harmony/main.go [hunk 5]
```diff
@@ -201,44 +217,51 @@ func createGlobalConfig() *nodeconfig.ConfigType {
...
+			panic(err)
+		}
+	} else {
+		nodeConfig = nodeconfig.GetShardConfig(uint32(*shardID))
 	}
 
 	nodeConfig.SelfPeer = p2p.Peer{IP: *ip, Port: *port, ConsensusPubKey: nodeConfig.ConsensusPubKey}
 
-	if accountIndex < core.GenesisShardNum { // The first node in a shard is the leader at genesis
+	if accountIndex < core.GenesisShardNum && !*isExplorer { // The first node in a shard is the leader at genesis
```
Comment (LeoHChen) [inline cmd/harmony/main.go]: is there a StringRole for explorer?

File: consensus/consensus_v2.go [hunk 6]
```diff
@@ -617,7 +617,7 @@ func (consensus *Consensus) finalizeCommits() {
 	// In current code, we add signatures in block in tryCatchup, the block dump to explorer does not contains signatures
 	// but since explorer doesn't need signatures, it should be fine
 	// in future, we will move signatures to next block
-	explorer.GetStorageInstance(consensus.leader.IP, consensus.leader.Port, true).Dump(block, beforeCatchupViewID)
+	explorer.GetStorageInstance(consensus.leader.IP, consensus.leader.Port, true).Dump(block, beforeCatchupNum)
```
Comment (LeoHChen) [inline consensus/consensus_v2.go]: why change this one?

Comment (rlan35) [reply consensus/consensus_v2.go]: This was wrong. It should be block number not view ID for identifying blocks.

File: cmd/harmony/main.go (hunk 5 above)
Comment (rlan35) [reply cmd/harmony/main.go]: Yes, I added a new role for explorer.

File: test/deploy.sh [hunk 7]
```diff
@@ -161,6 +161,7 @@ while IFS='' read -r line || [[ -n "$line" ]]; do
   esac
   case "${mode}" in leader*) args=("${args[@]}" -is_leader);; esac
   case "${mode}" in *archival|archival) args=("${args[@]}" -is_archival);; esac
+  case "${mode}" in explorer*) args=("${args[@]}" -is_genesis=false -is_explorer=true -shard_id=0);; esac
```
[the remark this answers is not in the document]
Comment (rlan35) [reply test/deploy.sh]: This is only for local test.

**Pull request merged**